crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -101,18 +101,33 @@ impl CheckpointStore {
         contents: CheckpointContents,
         epoch_store: &AuthorityPerEpochStore,
     ) {
-        if epoch_store.epoch() == checkpoint.summary.epoch {
-            epoch_store
-                .put_genesis_checkpoint_in_builder(&checkpoint.summary, &contents)
-                .unwrap();
-        } else {
-            debug!("Not inserting checkpoint builder data for genesis checkpoint, validator epoch {}, genesis epoch {}",
-                epoch_store.epoch(), checkpoint.summary.epoch,
-            );
+        assert_eq!(
+            checkpoint.epoch(),
+            0,
+            "can't call insert_genesis_checkpoint with a checkpoint not in epoch 0"
+        );
+
+        // Only insert the genesis checkpoint if the DB is empty and doesn't have it already
+        if self
+            .get_checkpoint_by_digest(&checkpoint.digest())
+            .unwrap()
+            .is_none()
+        {
+            if epoch_store.epoch() == checkpoint.summary.epoch {
+                epoch_store
+                    .put_genesis_checkpoint_in_builder(&checkpoint.summary, &contents)
+                    .unwrap();
+            } else {
+                debug!(
+                    validator_epoch =% epoch_store.epoch(),
+                    genesis_epoch =% checkpoint.epoch(),
+                    "Not inserting checkpoint builder data for genesis checkpoint",
+                );
+            }
+            self.insert_verified_checkpoint(checkpoint.clone()).unwrap();
+            self.insert_checkpoint_contents(contents).unwrap();
```

Comment:
Do we want those two calls swapped in places? Otherwise we insert verified_checkpoint without contents and if we crash between the calls we will stay in this state?

Reply:
done.